modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/LeaseTracker.java

```diff
@@ -0,0 +1,149 @@
+import org.apache.ignite.internal.metastorage.EntryEvent;
+import org.apache.ignite.internal.metastorage.MetaStorageManager;
+import org.apache.ignite.internal.metastorage.WatchEvent;
+import org.apache.ignite.internal.metastorage.WatchListener;
+import org.apache.ignite.internal.replicator.ReplicationGroupId;
+import org.apache.ignite.internal.table.distributed.replicator.TablePartitionId;
+import org.apache.ignite.internal.util.ByteUtils;
+import org.apache.ignite.internal.util.Cursor;
+import org.apache.ignite.internal.vault.VaultEntry;
+import org.apache.ignite.internal.vault.VaultManager;
+import org.apache.ignite.lang.ByteArray;
+
+/**
+ * Class tracks cluster leases in memory.
+ * At first, the class state recoveries from Vault, then updates on watch's listener.
+ */
+public class LeaseTracker {
+    /** Ignite logger. */
+    private static final IgniteLogger LOG = Loggers.forClass(LeaseTracker.class);
+
+    /** Vault manager. */
+    private final VaultManager vaultManager;
+
+    /** Metastorage manager. */
+    private final MetaStorageManager msManager;
+
+    /** Leases cache. */
+    private final Map<ReplicationGroupId, Lease> leases;
+
+    /** Listener to update a leases cache. */
+    private final UpdateListener updateListener = new UpdateListener();
+
+    /**
+     * The constructor.
+     *
+     * @param vaultManager Vault manager.
+     * @param msManager Metastorage manager.
+     */
+    public LeaseTracker(VaultManager vaultManager, MetaStorageManager msManager) {
+        this.vaultManager = vaultManager;
+        this.msManager = msManager;
+
+        this.leases = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Recoveries state from Vault and subscribers on further updates.
+     */
+    public void startTrack() {
+        msManager.registerPrefixWatch(ByteArray.fromString(PLACEMENTDRIVER_PREFIX), updateListener);
+
+        try (Cursor<VaultEntry> cursor = vaultManager.range(
+                ByteArray.fromString(PLACEMENTDRIVER_PREFIX),
+                ByteArray.fromString(incrementLastChar(PLACEMENTDRIVER_PREFIX))
+        )) {
+            for (VaultEntry entry : cursor) {
+                String key = entry.key().toString();
+
+                key = key.replace(PLACEMENTDRIVER_PREFIX, "");
+
+                TablePartitionId grpId = TablePartitionId.fromString(key);
+                Lease lease = ByteUtils.fromBytes(entry.value());
+
+                leases.put(grpId, lease);
+            }
+        }
+
+        LOG.info("Leases cache recovered [leases={}]", leases);
+    }
+
+    /**
+     * Stops the tracker.
+     */
+    public void stopTrack() {
+        msManager.unregisterWatch(updateListener);
+    }
+
+    /**
+     * Get a lease for a particular group.
+     *
+     * @param grpId Replication group id.
+     * @return A lease is associated with the group.
+     */
+    public Lease getLease(ReplicationGroupId grpId) {
+        return leases.getOrDefault(grpId, new Lease());
```

Comment:
this logic with getOrDefault is kind of unobvious.

Reply:
I can add @NotNull annotation, if it will more expected.